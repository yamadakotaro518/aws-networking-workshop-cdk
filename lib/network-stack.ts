import { Stack, StackProps } from "aws-cdk-lib";
import {
  SubnetConfiguration,
  SubnetType,
  Vpc
} from "aws-cdk-lib/aws-ec2";
import { Construct } from "constructs";

export interface NetworkProps extends StackProps {
  readonly vpcName: string;
  readonly cidr: string;
  readonly maxAzs?: number;
  readonly subnetConfiguration?: SubnetConfiguration[];
}
export const setDefaultValue = (props: NetworkProps) => {
  return {
    maxAzs: 2,
    subnetConfiguration: [
      {
        cidrMask: 24,
        name: "public",
        subnetType: SubnetType.PUBLIC,
      },
    ],
    ...props,
  };
};

/**
 * Network Stack
 * VPC及びVPC関連リソースを作成する
 */
export class NetworkStack extends Stack {
  readonly vpc: Vpc;

  constructor(scope: Construct, id: string, props: NetworkProps) {
    super(scope, id, props);

    this.vpc = new Vpc(this, "vpc", { ...setDefaultValue(props) });
  }
}
