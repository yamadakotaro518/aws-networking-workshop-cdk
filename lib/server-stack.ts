import { Stack, StackProps } from "aws-cdk-lib";
import {
  AmazonLinuxImage,
  IMachineImage,
  Instance,
  InstanceClass,
  InstanceSize,
  InstanceType,
  Peer,
  Port,
  SecurityGroup,
  SubnetSelection,
  SubnetType,
  Vpc
} from "aws-cdk-lib/aws-ec2";
import { ManagedPolicy, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
import { Construct } from "constructs";

export interface ServerProps extends StackProps {
  readonly vpc: Vpc;
  readonly instanceType?: InstanceType;
  readonly machineImage?: IMachineImage;
  readonly vpcSubnets?: SubnetSelection;
  /**
   * アクセス許可するIPアドレス
   */
  readonly allowCiderBlocks?: string[];
}
export const setDefaultValue = (props: ServerProps) => {
  return {
    instanceType: InstanceType.of(InstanceClass.T2, InstanceSize.MICRO),
    machineImage: new AmazonLinuxImage(),
    vpcSubnets: props.vpc.selectSubnets({ subnetType: SubnetType.PUBLIC }),
    ...props,
  };
};

/**
 * Server Stack
 * Server関連のリソースを作成する
 */
export class ServerStack extends Stack {
  readonly publicIpAdress: string;

  constructor(scope: Construct, id: string, props: ServerProps) {
    super(scope, id, props);

    const {
      vpc,
      instanceType,
      machineImage,
      vpcSubnets,
      allowCiderBlocks
    } = setDefaultValue(props);

    // セキュリティグループ作成
    const securityGroup = new SecurityGroup(this, 'securityGroup', {
      securityGroupName: 'security group of ec2',
      vpc: vpc
    })

    allowCiderBlocks?.forEach(ip => {
      securityGroup.addIngressRule(Peer.ipv4(ip), Port.tcp(22))
      securityGroup.addIngressRule(Peer.ipv4(ip), Port.icmpPing())
      securityGroup.addIngressRule(Peer.ipv4(ip), Port.tcp(5201))
    })

    // ec2用のロール作成
    const role = new Role(this, 'ec2role', {
      assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore')
      ]
    })

    const instance = new Instance(this, "ec2Instance", {
      vpc,
      instanceType,
      machineImage,
      vpcSubnets,
      securityGroup,
      role
    });
    this.publicIpAdress = instance.instancePublicIp
  }
}
