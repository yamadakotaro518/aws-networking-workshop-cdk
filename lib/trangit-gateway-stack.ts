import { Stack, StackProps } from 'aws-cdk-lib';
import { CfnTransitGateway, CfnTransitGatewayAttachment, Vpc, SubnetType, Subnet, CfnRoute } from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

export interface TransitGatewayProps extends StackProps {
  readonly vpcs: Vpc[];
}

/**
 * transit gateway stack
 */
export class TransitGatewayStack extends Stack {
  readonly transitGatewayId: string;

  constructor(scope: Construct, id: string, props: TransitGatewayProps) {
    super(scope, id, props);

    const transitGateway = new CfnTransitGateway(this, 'transitGateway', {
      description: 'Immersion Day TGW'
    })
    this.transitGatewayId = transitGateway.attrId;

    props.vpcs.forEach((vpc, vpcIndex) => {
      // transit gateway用のsubnetを2つ作成
      const subnet1ForTGW = new Subnet(this, `subnet1ForTGW-vpc${vpcIndex}`, {
        availabilityZone: 'ap-northeast-1a',
        vpcId: vpc.vpcId,
        cidrBlock: `10.${vpcIndex}.2.0/28` 
      })

      const subnet2ForTGW = new Subnet(this, `subnet2ForTGW-vpc${vpcIndex}`, {
        availabilityZone: 'ap-northeast-1c',
        vpcId: vpc.vpcId,
        cidrBlock: `10.${vpcIndex}.3.0/28` 
      })

      // transit gateway attachmentを作成
      new CfnTransitGatewayAttachment(this, `transitGatewayAttachment-vpc${vpcIndex}`, {
        transitGatewayId: transitGateway.attrId,
        vpcId: vpc.vpcId,
        subnetIds: [subnet1ForTGW.subnetId, subnet2ForTGW.subnetId]
      })

      // サブネットのルートテーブルの向き先にTGWを追加
      vpc.publicSubnets.forEach((publicSubnet, subnetIndex) => {
        new CfnRoute(this, `routeForTGW-vpc${vpcIndex}-subnet${subnetIndex}`,{
          routeTableId: publicSubnet.routeTable.routeTableId,
          destinationCidrBlock: '10.0.0.0/8',
          transitGatewayId: this.transitGatewayId
        })
      })
    })
  }
}