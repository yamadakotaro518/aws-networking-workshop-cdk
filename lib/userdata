#cloud-config

runcmd:
  - sudo yum install iperf3 -y