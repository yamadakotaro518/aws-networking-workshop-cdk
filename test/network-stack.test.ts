import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { SubnetType } from 'aws-cdk-lib/aws-ec2';
import { NetworkStack } from '../lib/network-stack';

const env = {
  account: 'XXXXXXXXXXXX',
  region: 'ap-northeast-1'
}

test('network stack', () => {
  const app = new App();
  const stack = new NetworkStack(app, 'network', {
    env,
    vpcProps: {
      vpcName: 'VPC A',
      cidr: '10.0.0.0/16',
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'VPC A - AZ1',
          subnetType: SubnetType.PUBLIC
        }
      ]
    }
  })
  expect(Template.fromStack(stack).toJSON()).toMatchSnapshot(); 
})