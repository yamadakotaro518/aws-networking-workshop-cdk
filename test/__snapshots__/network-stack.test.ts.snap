// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`network stack 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "vpcA2121C38": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC A",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "vpcIGWE57CBDCA": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPC A",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "vpcVPCAAZ1Subnet1DefaultRoute58DB7E06": Object {
      "DependsOn": Array [
        "vpcVPCGW7984C166",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "vpcIGWE57CBDCA",
        },
        "RouteTableId": Object {
          "Ref": "vpcVPCAAZ1Subnet1RouteTableC9B641A0",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcVPCAAZ1Subnet1RouteTableAssociationFD5BF6C2": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcVPCAAZ1Subnet1RouteTableC9B641A0",
        },
        "SubnetId": Object {
          "Ref": "vpcVPCAAZ1Subnet1Subnet4C31D41E",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcVPCAAZ1Subnet1RouteTableC9B641A0": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "network/vpc/VPC A - AZ1Subnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcVPCAAZ1Subnet1Subnet4C31D41E": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.0.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "VPC A - AZ1",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "network/vpc/VPC A - AZ1Subnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcVPCAAZ1Subnet2DefaultRouteE60746E3": Object {
      "DependsOn": Array [
        "vpcVPCGW7984C166",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "vpcIGWE57CBDCA",
        },
        "RouteTableId": Object {
          "Ref": "vpcVPCAAZ1Subnet2RouteTable9A35EE97",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcVPCAAZ1Subnet2RouteTable9A35EE97": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "network/vpc/VPC A - AZ1Subnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcVPCAAZ1Subnet2RouteTableAssociationEE052844": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcVPCAAZ1Subnet2RouteTable9A35EE97",
        },
        "SubnetId": Object {
          "Ref": "vpcVPCAAZ1Subnet2Subnet7F2EBDBE",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcVPCAAZ1Subnet2Subnet7F2EBDBE": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1b",
        "CidrBlock": "10.0.1.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "VPC A - AZ1",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "network/vpc/VPC A - AZ1Subnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcVPCAAZ1Subnet3DefaultRoute1AC8C662": Object {
      "DependsOn": Array [
        "vpcVPCGW7984C166",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "vpcIGWE57CBDCA",
        },
        "RouteTableId": Object {
          "Ref": "vpcVPCAAZ1Subnet3RouteTable42D88D29",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcVPCAAZ1Subnet3RouteTable42D88D29": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "network/vpc/VPC A - AZ1Subnet3",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcVPCAAZ1Subnet3RouteTableAssociation2B01E091": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcVPCAAZ1Subnet3RouteTable42D88D29",
        },
        "SubnetId": Object {
          "Ref": "vpcVPCAAZ1Subnet3Subnet33613895",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcVPCAAZ1Subnet3Subnet33613895": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1c",
        "CidrBlock": "10.0.2.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "VPC A - AZ1",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "network/vpc/VPC A - AZ1Subnet3",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcVPCGW7984C166": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "vpcIGWE57CBDCA",
        },
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
